More noobly

noobanidus opened the pull request:
Mostly entity-related stuff, institutes a new (basically copied) IMultiPartEntity, provides a proper abstract template (basically just a copy of the 1.12 MultiPartEntity).

noobanidus committed:
Simplify a check.

noobanidus committed:
Comment out some GuiHandler things; fix Advancement function.

noobanidus committed:
These were already moved.

noobanidus committed:
Unused function.

noobanidus committed:
Move projectiles into their own directory.

noobanidus committed:
Overhauld StructureWoodVariant.

It should now be used for whatever purpose it needs to be used for.

noobanidus committed:
Begin work on the multipart entities. onUpdate -> tick

noobanidus committed:
Tweaks to particles, explosions, vectors.

noobanidus committed:
Hydra hydra hydra! Hydra kids, hydra wife.

noobanidus committed:
Finalise work on Bosses (naga, Ice Waifu)

noobanidus committed:
A bunch more commits.

noobanidus committed:
Projectiles.

noobanidus committed:
Merge branch '1.15.x' into more_noobly

noobanidus committed:
Run overrides inspection.

Andromander commented on src/main/java/twilightforest/entity/IEntityMultiPart.java:
```diff
@@ -0,0 +1,14 @@
+import net.minecraft.world.World;
+
+public interface IEntityMultiPart
+{
```
Bracket's on a new line rather than same line.

noobanidus replied:
Whoops, I literally copy/pasted from Vanilla.

Andromander commented on src/main/java/twilightforest/entity/IEntityMultiPart.java:
```diff
@@ -0,0 +1,14 @@
+
+    boolean attackEntityFromPart(MultiPartEntityPart part, DamageSource source, float damage);
+
+	Entity[] getParts();
```
Indent here looks a little off compared to the others.

Andromander commented on src/main/java/twilightforest/entity/boss/EntityTFHydra.java:
```diff
@@ -724,7 +728,7 @@ public ResourceLocation getLootTable() {
 
 	@Override
-	protected boolean canDespawn() {
+	public boolean canDespawn(double IDontKnowWhatThisIs) {
```
I think last time I looked at this, the parameter was used to check distance of the nearest player.

williewillus replied:
yes, it's `distance`

noobanidus replied:
Awesome, thanks.

Andromander reviewed (approved):
Mostly just little things. Otherwise, should be good to go.

williewillus commented on src/main/java/twilightforest/entity/EntityTFChainBlock.java:
```diff
@@ -244,6 +245,11 @@ public void tick() {
 
+	@Override
+	protected void registerData() {
+		// TODO: What does this do?
```
for registering datawatcher entries, which this entity has none. the todo can be removed

noobanidus replied:
In that case, I'm going to move into the abstract base class and it can be overriden if needed.

noobanidus replied:
Actually, scratch that, that might be unclear. I'll just remove the todo.

williewillus commented on src/main/java/twilightforest/entity/EntityTFIceExploder.java:
```diff
@@ -79,8 +81,9 @@ protected void onDeathUpdate() {
 		if (this.deathTime == 60) // delay until 3 seconds
 		{
 			if (!world.isRemote) {
+				// TODO: I think this is correct
```
not fully equivalent to the old parameter, but good enough. todo can be removed.

williewillus commented on src/main/java/twilightforest/entity/EntityTFSlideBlock.java:
```diff
@@ -101,10 +100,11 @@ protected void registerData() {
 	}
 
-	@Override
+	// TODO: Still don't know what this is now
```
bypassesSteppingEffects

williewillus commented on src/main/java/twilightforest/entity/EntityTFSlideBlock.java:
```diff
@@ -158,20 +158,19 @@ public void tick() {
-//					this.motionZ *= 0.699999988079071D;
-//					this.motionY *= 0.699999988079071D;
-					this.getMotion().mul(0.699999988079071D, 0.699999988079071D, 0.699999988079071D);
+					// TODO: mul just returns a new Vec3d, it doesn't actually motify the speed
```
this seems correct? todo can be removed.

williewillus commented on src/main/java/twilightforest/entity/EntityTFSlideBlock.java:
```diff
@@ -158,20 +158,19 @@ public void tick() {
 					}
 				} else if (this.slideTime > 100 && (pos.getY() < 1 || pos.getY() > 256) || this.slideTime > 600) {
-					this.entityDropItem(new ItemStack(this.myState.getBlock(), 1, this.myState.getBlock().damageDropped(this.myState)), 0.0F);
+					this.entityDropItem(new ItemStack(this.myState.getBlock()), 0.0F);
```
it should probably query the loot table.

noobanidus replied:
As it appears to be a "falling block" kind of block (except it's a sliding block), I wasn't sure if that value would be determinable inside of the loot table -- it's predicated in the constructor on line 48.

noobanidus replied:
Unless you meant the block loot table...

williewillus commented on src/main/java/twilightforest/entity/EntityTFSlideBlock.java:
```diff
@@ -246,6 +245,12 @@ public boolean isPushedByWater() {
+	@Override
+	public IPacket<?> createSpawnPacket() {
+		// TODO: This needs to be done
+		return null;
```
NetworksHooks.getEntitySpawningPacket or something like that.

williewillus commented on src/main/java/twilightforest/entity/boss/EntityTFIceCrystal.java:
```diff
@@ -19,9 +20,14 @@
 	public EntityTFIceCrystal(EntityType<? extends EntityTFIceCrystal> type, World world) {
 		super(type, world);
 
+		// TODO: ???
```
if it was already commented in 1.12, remove it. otherwise it's just a setHeldItem call

noobanidus replied:
Ah, yeah, blame says it's been commented out for five years.

williewillus commented on src/main/java/twilightforest/entity/boss/EntityTFNaga.java:
```diff
@@ -446,10 +445,11 @@ private void setSegmentsPerHealth() {
 
-	@Override
+	// TODO: I don't know what this is now
+/*	@Override
```
bypassesSteppingEffects

williewillus commented on src/main/java/twilightforest/entity/passive/EntityTFMobileFirefly.java:
```diff
@@ -97,10 +99,11 @@ protected void updateAITasks() {
-//		return false;
-//	}
+	// TODO: Can't actually find what this is now
+/*	@Override
```
bypassesSteppingEffects

williewillus commented on src/main/java/twilightforest/entity/passive/EntityTFMobileFirefly.java:
```diff
@@ -138,6 +142,12 @@ public float getGlowBrightness() {
 		return (float) Math.sin(this.ticksExisted / 7.0) + 1F;
 	}
 
+	// TODO: I'm unsure if this is sufficient?
```
should be

williewillus commented on src/main/java/twilightforest/entity/projectile/EntityTFIceSnowball.java:
```diff
@@ -25,6 +25,11 @@ public EntityTFIceSnowball(EntityType<? extends EntityTFIceSnowball> type, World
 
+	@Override
+	protected void registerData() {
+		// TODO: Is this needed?
```
leave blank

williewillus commented on src/main/java/twilightforest/entity/projectile/EntityTFMoonwormShot.java:
```diff
@@ -27,6 +30,11 @@ public EntityTFMoonwormShot(EntityType<? extends EntityTFMoonwormShot> type, Wor
 
+	@Override
+	protected void registerData() {
+		// TODO: Is this needed?
```
same

williewillus commented on src/main/java/twilightforest/entity/projectile/EntityTFMoonwormShot.java:
```diff
@@ -73,7 +81,9 @@ public void handleStatusUpdate(byte id) {
-				this.world.addParticle(ParticleTypes.BLOCK_CRACK, this.getX(), this.getY(), this.getZ(), 0.0D, 0.0D, 0.0D, stateId);
+				// TODO: Don't think this is enough
+				// state ID needs to be retained somehow
+				this.world.addParticle(ParticleTypes.DAMAGE_INDICATOR, false, this.getX(), this.getY(), this.getZ(), 0.0D, 0.0D, 0.0D);
```
this is the wrong particle type, this is the hearts that appear when you attack something.
should be something like `new BlockParticleData(ParticleTypes.BLOCK, state)`

there was also another place in this PR where you used damage_indicator that needs to be fixed too.

noobanidus replied:
Thanks, I knew this wasn't right.

williewillus commented on src/main/java/twilightforest/entity/projectile/EntityTFNatureBolt.java:
```diff
@@ -33,6 +33,11 @@ public EntityTFNatureBolt(EntityType<? extends EntityTFNatureBolt> type, World w
 
+	@Override
+	protected void registerData() {
+		// TODO: Needed?
```
nope

williewillus commented on src/main/java/twilightforest/entity/projectile/EntityTFNatureBolt.java:
```diff
@@ -59,7 +64,8 @@ public void handleStatusUpdate(byte id) {
 			for (int i = 0; i < 8; ++i) {
-				this.world.addParticle(ParticleTypes.BLOCK_CRACK, this.getX(), this.getY(), this.getZ(), rand.nextGaussian() * 0.05D, rand.nextDouble() * 0.2D, rand.nextGaussian() * 0.05D, stateId);
+				// TODO: Pretty sure that this is needed to be modified
+				this.world.addParticle(ParticleTypes.DAMAGE_INDICATOR, false, this.getX(), this.getY(), this.getZ(), rand.nextGaussian() * 0.05D, rand.nextDouble() * 0.2D, rand.nextGaussian() * 0.05D);
```
BlockParticleData as mentioned above

williewillus commented on src/main/java/twilightforest/entity/projectile/EntityTFTomeBolt.java:
```diff
@@ -27,6 +27,11 @@ public EntityTFTomeBolt(EntityType<? extends EntityTFTomeBolt> type, World world
 
+	@Override
+	protected void registerData() {
+		// TODO: Needed?
```
nope

williewillus commented on src/main/java/twilightforest/entity/projectile/EntityTFTwilightWandBolt.java:
```diff
@@ -32,6 +32,11 @@ public EntityTFTwilightWandBolt(EntityType<? extends EntityTFTwilightWandBolt> t
 
+	@Override
+	protected void registerData() {
+		// TODO: Needed?
```
same

williewillus commented on src/main/java/twilightforest/enums/StructureWoodVariant.java:
```diff
@@ -32,40 +36,29 @@
+        this.gate       = blocks[5].get(); // gate  ;
+        this.plate      = blocks[6].get(); // plate ;
+        this.blocks     = Stream.of(blocks).map(Supplier::get).collect(Collectors.toList());
     }
```
why?
passing them explicitly is cleaner than an array whose contents are completely up to convention
at least with parameters we have the param names.

noobanidus replied:
this is true, I'll just explicitly create an array instead

noobanidus replied:
Reverted and changed to an explicit array creation.

williewillus commented on src/main/java/twilightforest/enums/StructureWoodVariant.java:
```diff
@@ -81,76 +74,35 @@ public String getName() {
+    }
 
-        if (!"minecraft".equals(block.getRegistryName().getNamespace())) return null;
+    // Redone to make more sense with the above
```
will this comment make sense 6 months from now? (redone w.r.t what?)
I'd just remove it.

noobanidus replied:
Very true, removed.

williewillus reviewed (changes requested):
some stuff. thanks for helping us trawl through this.

noobanidus committed:
Remove unneeded TODOs.

noobanidus committed:
canTriggerWalking -> bypassesSteppingEffect

noobanidus committed:
Fix indentation issues.

noobanidus committed:
Properly instantiate network packet for SlideBlock.

noobanidus committed:
Correctly name parameter.

noobanidus committed:
Reindent.

noobanidus committed:
Actually remove that network hook.

noobanidus committed:
Remove unused code from 5+ years ago.

noobanidus committed:
Fix additional motion (getMotion().add returns a new Vec3d, not modify-in-place)

noobanidus committed:
Remove unneeded TODO.

noobanidus committed:
Update to properly use ItemParticleData.

In this instance, as it requires an ItemStack, I've made a static field
with an Ender Pearl ItemStack in it on the client side to prevent
constant re-instantiation.

I'm not sure how much of an overhead saving this is but it feels better.

noobanidus committed:
Correctly usue BlockParticleData.

noobanidus committed:
Another BlockParticleData.

noobanidus committed:
Correct Item/BlockParticleData.

noobanidus committed:
Reindent and revert change to varargs.

noobanidus committed:
Remove needless comment.

Drullkus reviewed (approved):
Aside from Willie's requested changes, it looks good to me. Mergable after Willie approves

noobanidus commented:
Yeah, there's on thing I left unmarked, too, re: loot tables, but as it appeared to be a falling entity I wasn't sure how it should be handled.

williewillus commented:
good enough for now
